libimage/pull.go

```diff
@@ -309,19 +307,30 @@ func (r *Runtime) copyFromDefault(ctx context.Context, ref types.ImageReference,
 		// instead of looking at the StringWithinTransport().
 		storageName, err = getImageID(ctx, ref, nil)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
 		imageName = "sha256:" + storageName[1:]
 	}
 
 	// Create a storage reference.
 	destRef, err := storageTransport.Transport.ParseStoreReference(r.store, storageName)
 	if err != nil {
-		return nil, fmt.Errorf("parsing %q: %w", storageName, err)
+		return nil, nil, fmt.Errorf("parsing %q: %w", storageName, err)
 	}
 
 	_, err = c.Copy(ctx, ref, destRef)
-	return []string{imageName}, err
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to perform copy: %w", err)
+	}
+	if resolvedReference == nil { // resolvedReference should always be set for storageTransport destinations
```

Comment:
There are now three copies of this logic setting `resolvedReference` and looking up the image.

Can that be consolidated? One possible way to do this would be to have a `copier.copyToStorage(ctx, srcRef, destRef)` directly returning the resolved image; `copier.Copy` would then be a thin wrapper, with both `Copy` and `copyToStorage` calling `copier.copy(ctx, src, dest, reportResolvedReference)` containing most of the implementation.